### PR TITLE
#582 update drug registration template

### PR DESCRIPTION
**Pull request by nmadruga**
Fixes #582

### Changes
- Update codes to be more descriptive
- Remove `ids` from each question
- Add `dropdownChoice` with **search** and list fields
- Add `dropdownChoice` to present result from GraphQL API calls to UC and other server to get country names
- Change to `radioChoice` questions with less options to select, using "Other" new functionality
- Add `longText` for final comment

### CODE REVIEW

**Comment by CarlosNZ** [comment]
Hey, I'm getting this error when running `insertData`:
```
  -- 03_template_E_drugRegistration.js
[
  {
    "message": "duplicate key value violates unique constraint \"template_stage_pkey\"",
    "locations": [
      {
        "line": 2,
        "column": 5
      }
    ],
    "path": [
      "createTemplate"
    ]
  }
]
```

I had this exact same problem recently, it's to do with the ID auto-incrementer getting messed up by manual ID insertion. Let me know if you need a hand to sort it, cos it can be super-annoying.

**Comment by CarlosNZ** [comment]
> I had this exact same problem recently, it's to do with the ID auto-incrementer getting messed up by manual ID insertion. Let me know if you need a hand to sort it, cos it can be super-annoying.

Okay, yes, it's the same thing I had -- you can fix it by removing the IDs from `templateStagesUsingId` in the "Basic Review Testing" template. That's what's messing up the serial counter.

File: database/insertData/03_template_E_drugRegistration.js [hunk 1]
```diff
@@ -406,7 +781,7 @@ exports.queries = [
                       category: INFORMATION
                     }
                     {
-                      code: "S1Info7"
+                      code: "S1TextPage6"
```
Comment (CarlosNZ) [inline database/insertData/03_template_E_drugRegistration.js:784]: The "Shelf life" question (below) -- maybe add the "must be a number" validation to it?

File: database/insertData/03_template_E_drugRegistration.js [hunk 2]
```diff
@@ -474,14 +848,13 @@ exports.queries = [
...
               {
-                id: 1009
                 code: "S3"
                 title: "Product images"
                 index: 2
                 templateElementsUsingId: {
                   create: [
                     {
-                      code: "S3Q1"
+                      code: "S3Q1UploadSamples"
```
Comment (CarlosNZ) [inline database/insertData/03_template_E_drugRegistration.js:857]: This question missing `fileCountLimit` even though you've mentioned it in description. ;)

File: database/insertData/03_template_E_drugRegistration.js [hunk 3]
```diff
@@ -541,7 +912,6 @@ exports.queries = [
                 }
               }
               {
-                id: 10
```
Comment (CarlosNZ) [inline database/insertData/03_template_E_drugRegistration.js:544]: Can you add `${devActions}` just below `${coreActions}` and import it at the top? (see other templates)

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Awesome, great to finally see all the bells and whistles in action :) 

Couple of additional comments:
- should we make `isLinear: true`? I know non-linear is easier for development purposes, but the "proper" drug reg would be linear, right? Maybe we leave it for now, but update it later.
- Is the `templateStageReviewLevelsUsingId` correct, or are you just leaving it for now? (currently 1 review level per stage, so no consolidation, and they're all called "Review")
- Might be a good idea to add colours to template stages (see Basic Review Testing)

**Comment by nmadruga** [comment]
> templateStagesUsingId

Thanks! Done: bcbacfb

**Comment by nmadruga** [comment]
> * should we make `isLinear: true`? I know non-linear is easier for development purposes, but the "proper" drug reg would be linear, right? Maybe we leave it for now, but update it later.

Yes, I didn't see a actual need for it to be Linear now, so let's leave like that and changing that will be quite easy, but probably annoying for them to check out and test it as well.

**Comment by nmadruga** [comment]
> * Is the `templateStageReviewLevelsUsingId` correct, or are you just leaving it for now? (currently 1 review level per stage, so no consolidation, and they're all called "Review")

I think that until we have a Consolidation in place we should just go this far, so I didn't change that at all

**Comment by nmadruga** [comment]
> * Might be a good idea to add colours to template stages (see Basic Review Testing)

Cool, that's a great idea - so the list will be using the same colours. Done: fa98bd9

File: database/insertData/03_template_E_drugRegistration.js [hunk 4]
```diff
@@ -22,41 +22,123 @@ exports.queries = [
...
-                      code: "S1Q2"
+                      code: "S1Q2Origin"
                       index: 20
                       title: "Origin category"
                       elementTypePluginCode: "radioChoice"
                       category: QUESTION
                       parameters: {
                         label: "Product origin"
-                        description: "_Select which is the origin of the drug._"
+                        description: "_Select which is the origin of the product._"
```
[the remark this answers is not in the document]
Comment (nmadruga) [reply database/insertData/03_template_E_drugRegistration.js:83]: Have you added a new issue for this one yet?

File: database/insertData/03_template_E_drugRegistration.js [hunk 5]
```diff
@@ -142,24 +268,39 @@ exports.queries = [
...
+                          { code: "LCD", name: "Limited Controlled Medicine" },
+                          { code: "OTC", name: "Over The Counter Medicine" },
+                          { code: "POM", name: "Prescription Only Medicine" }
                         ]
+                        optionsDisplayProperty: "name"
                       }
                     }
                     {
-                      code: "S1Q9"
+                      code: "S1Q9Strength"
```
[the remark this answers is not in the document]
Comment (nmadruga) [reply database/insertData/03_template_E_drugRegistration.js:303]: I don't know exactly what it is either : (

File: database/insertData/03_template_E_drugRegistration.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply database/insertData/03_template_E_drugRegistration.js:83]: Done [#645](https://github.com/openmsupply/application-manager-web-app/issues/645) -- haven't added an Epic label yet, feel free to pick one you think's appropriate.

File: database/insertData/03_template_E_drugRegistration.js (hunk 3 above)
Comment (nmadruga) [reply database/insertData/03_template_E_drugRegistration.js:544]: Done: ecfbfcb

**Comment by nmadruga** [comment]
Thanks for all suggestions Carl, I've addressed each one (that doesn't have a comment on it) in this commit: 351efd6

Sorry I went `resolving` them all to be quicker for me to see what was remaining to do. Trying to make flash review now : D

File: database/insertData/03_template_E_drugRegistration.js [hunk 6]
```diff
@@ -22,41 +22,123 @@ exports.queries = [
...
+                              {
+                                operator: "objectProperties"
+                                children: ["responses.Q1.text"]
+                              }
+                              { value: "" }
+                            ]
+                          }
+                        ]
+                      }
+                      validationMessage: "You need a valid product name."
```
[the remark this answers is not in the document]
Comment (nmadruga) [reply database/insertData/03_template_E_drugRegistration.js]: Yes, I did this one first and them haven't updated. I tried using Regex, but it wasn't working. So for now just checking for non-empty string. Done: dccfce7

[the remark this answers is not in the document]
Comment (CarlosNZ) [reply database/insertData/03_template_E_drugRegistration.js]: Why do you need any validation at all? If it's required, it'll already be checked for `null` and empty string.

**Review by CarlosNZ — APPROVED** [review]
Great, thanks a lot.

I'm still not entirely clear why you need any "not-empty" validation on a required question (as per [this question](https://github.com/openmsupply/application-manager-server/pull/330#discussion_r623489056), but it's fine either way.

**Comment by nmadruga** [comment]
> Great, thanks a lot.
> 
> I'm still not entirely clear why you need any "not-empty" validation on a required question (as per [this question](https://github.com/openmsupply/application-manager-server/pull/330#discussion_r623489056), but it's fine either way.

That's a good point. I tried to use a REGEX to only allow a Product with some valid name - not allowing only white spaces for example. But after a few attempts this was not working as I wanted, so left that not-so-useful validation there

**Pull request merged**